guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+export function isFile() {
+  // ...
+}
+
+export function isDirectory() {
+  // ...
+}
+
+// file-utils.test.ts
+
+describe('isFile', () => {
+  // ...
+});
+
+describe('isDirectory', () => {
+  // ...
+});
+```
+
+✅
+
+``` typescript
+// file-utils.ts
+
+export function isFile() {
+  // ...
+}
+
+export function isDirectory() {
+  // ...
+}
+
+// file-utils.test.ts
+
+describe('file-utils', () => {
+  describe('isFile', () => {
+    // ...
+  });
+
+  describe('isDirectory', () => {
+    // ...
+  });
+});
+```
+
+## Use `describe` to group tests for methods
+
+Using `describe` to wrap tests for a method makes it easier to spot these tests in a large file and makes it possible to run them on their own (using `.only`). It also helps to establish the subject of the test (the "it") and therefore keeps the test well named and focused.
+
+🚫
+
+``` typescript
+describe('KeyringController', () => {
+  it('adds a new account to the given keyring', () => {
+    // ...
+  });
+
+  it('removes the account identified by the given address from its associated keyring', () => {
+    // ...
+  });
+});
+```
+
+✅
+
+``` typescript
+describe('KeyringController', () => {
+  describe('addAccount', () => {
+    it('adds a new account to the given keyring', () => {
+      // ...
+    });
+  });
+
+  describe('removeAccount', () => {
+    it('removes the account identified by the given address from its associated keyring', () => {
+      // ...
+    });
+  });
+});
+```
+
+## 💡 Use `describe` to group tests for scenarios
+
+If you have many tests that verify the behavior of a piece of code under a particular condition, you might find it helpful to wrap those tests with a `describe` block. Use a phrase starting with `if ...` or `when ...` so as to form a sentence when combined with the test name.
+
+1️⃣
+
+``` typescript
+it('calls addToken on the tokens controller if "addToken" is true', () => {
+  // ...
+})
+
+it('adds the new token to the state if "addToken" is true', () => {
+  // ...
+})
+
+it('returns a promise that resolves to true if "addToken" is true', () => {
+  // ...
+})
+```
+
+2️⃣
+
+``` typescript
+describe('if "addToken" is true', () => {
+  it('calls addToken on the tokens controller', () => {
+    // ...
+  });
+
+  it('adds the new token to the state', () => {
+    // ...
+  });
+
+  it('returns a promise that resolves to true', () => {
+    // ...
+  });
+});
+```
+
+## Use `it` to specify known behavior for a unit of code
+
+Use `it` to describe the behavior of a piece of code from the consumer's perspective as it stands today, not as it ought to be.
+
+Do not use "should" at the start of the test name. The official Jest documentation [omits this word from their examples](https://jestjs.io/docs/next/getting-started), and it creates noise when reviewing the list of tests that Jest outputs after a run.
+
+Do not repeat the name of the function or method in the name of the test if it is possible to communicate what a consumer expects more clearly.
```

Comment:
Is the qualifier necessary here? e.g. can you think of an example where it would be difficult to avoid repeating the name of the function?
```suggestion
Do not repeat the name of the function or method in the name of the test.
```

Reply:
By adding this clause I was trying to admit that naming is hard, and sometimes the method already describes what it does well and it's really hard to come up with other words. I guess that's not really what I'm saying though. I can remove this.